pyt/expr_visitor.py

```diff
@@ -537,6 +539,108 @@ def process_function(self, call_node, definition):
+
+        # Create e.g. ¤call_1 = ret_func_foo
+        LHS = CALL_IDENTIFIER + 'call_' + str(saved_function_call_index)
+        RHS = 'ret_' + call_label.result[:index] + '('
+
+        call_node = BBorBInode(
+            label='',
+            left_hand_side=LHS,
+            right_hand_side_variables=[],
+            line_number=node.lineno,
+            path=self.filenames[-1],
+            func_name=call_label.result[:index]
+        )
+        visual_args = list()
+        rhs_vars = list()
+        last_return_value_of_nested_call = None
+
+        for arg in itertools.chain(node.args, node.keywords):
+            if isinstance(arg, ast.Call):
+                return_value_of_nested_call = self.visit(arg)
+
+                if last_return_value_of_nested_call:
+                    # connect inner to other_inner in e.g.
+                    # `scrypt.outer(scrypt.inner(image_name), scrypt.other_inner(image_name))`
+                    # I should probably loop to the inner most call of other_inner here.
+                    try:
+                        last_return_value_of_nested_call.connect(return_value_of_nested_call.first_node)
+                    except AttributeError:
+                        last_return_value_of_nested_call.connect(return_value_of_nested_call)
+                else:
+                    # I should only set this once per loop, inner in e.g.
+                    # `scrypt.outer(scrypt.inner(image_name), scrypt.other_inner(image_name))`
+                    # (inner_most_call is used when predecessor is a ControlFlowNode in connect_control_flow_node)
+                    call_node.inner_most_call = return_value_of_nested_call
+                last_return_value_of_nested_call = return_value_of_nested_call
+
+                visual_args.append(return_value_of_nested_call.left_hand_side)
+                rhs_vars.append(return_value_of_nested_call.left_hand_side)
+            else:
+                label = LabelVisitor()
+                label.visit(arg)
+                visual_args.append(label.result)
+
+                vv = VarsVisitor()
+                vv.visit(arg)
+                rhs_vars.extend(vv.result)
+        if last_return_value_of_nested_call:
+            # connect other_inner to outer in e.g.
+            # `scrypt.outer(scrypt.inner(image_name), scrypt.other_inner(image_name))`
+            last_return_value_of_nested_call.connect(call_node)
+
+        if len(visual_args) > 0:
+            for arg in visual_args:
+                RHS = RHS + arg + ", "
+            # Replace the last ", " with a )
+            RHS = RHS[:len(RHS) - 2] + ')'
+        else:
+            RHS = RHS + ')'
+        call_node.label = LHS + " = " + RHS
+
+        # .args is only used in get_sink_args
+        # We make a new list because right_hand_side_variables is extended in assignment_call_node
+        call_node.right_hand_side_variables = rhs_vars
+        call_node.args = list(rhs_vars)
```

Comment:
If we did not make a new list, things like this would happen:
```python
 > User input at line 8, trigger word "request.args.get(":
     ¤call_1 = ret_request.args.get('image_name')
Reassigned in:
    File: example/vulnerable_code/path_traversal_sanitised.py
     > Line 8: image_name = ¤call_1
File: example/vulnerable_code/path_traversal_sanitised.py
 > reaches line 10, trigger word "replace(":
    ¤call_2 = ret_image_name.replace('..', '')
```
though the problem on master was that we didn't include `~call_N` in .args, and thus in the return value of `get_sink_args` and so we had a false-negative that's now fixed. However, `or_in_sink` remains unsolved, for the time being.

Reply:
The reason `or_in_sink.py` doesn't pass yet is because we don't `if isinstance` for `BoolOp` in the 'loop through args' code in `visit_Call`, should be easy enough.